File: Day334_2025/Q334_GFG.py
class Solution:
    def countSubs(self, s: str) -> int:
        n = len(s)
        dp = [0] * (n + 1)

        for i in range(1, n + 1):
            for j in range(i - 1, 0, -1):
                cnt = 0
                t = i

                while j >= 1 and s[t - 1] == s[j - 1]:
                    cnt += 1
                    t -= 1
                    j -= 1

                if cnt > 0:
                    j += 1
                
                dp[i] = max(dp[i], cnt)

        ans = 0
        for i in range(1, n + 1):
            ans += i - dp[i]

        return ans

File: Day334_2025/Q334_GFG.cpp
class Solution {
  public:
    int countSubs(string& str) {
        // Q. Count of distinct substrings
        // code here
        int n=str.length();
        vector<int>dp(n+1,0);
        for(int i=1;i<=n;i++){
            for(int j=i-1;j>=1;j--){
                int cnt=0;
                int t=i;
                while(j>=1 && str[t-1]==str[j-1]){
                    cnt++;
                    t--;
                    j--;
                }
                if(cnt){
                    j++;
                }
                dp[i]=max(dp[i],cnt);
            }
        }
        int ans=0;
        for(int i=1;i<=n;i++){
            ans+=i-dp[i];
        }
        return ans;
    }
};